optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/reproducer/DroolsReproducer.java

```diff
@@ -0,0 +1,312 @@
+        RuntimeException reproducedException = journal.replay(oldKieSession);
+        assertOriginalExceptionReproduced("Cannot reproduce original exception after pruning the journal. " +
+                "This is a bug!");
+        return reproducedException;
+    }
+
+    public DroolsReproducer(KieSession oldKieSession, RuntimeException originalException, KieSessionJournal journal) {
+        this.oldKieSession = oldKieSession;
+        this.originalException = originalException;
+        this.journal = journal;
+    }
+
+    private void dropOldestUpdates() {
+        log.info("Dropping oldest updates...", journal.getMoveOperations().size());
+        HeadCuttingMutator m = new HeadCuttingMutator(journal.getMoveOperations());
+        while (m.canMutate()) {
+            long start = System.currentTimeMillis();
+            KieSessionJournalImpl testJournal = new KieSessionJournalImpl(journal.getFacts(), journal.getInitialInserts(), m.mutate());
+            boolean reproduced = reproduce(testJournal);
+            double tookSeconds = (System.currentTimeMillis() - start) / 1000d;
+            String outcome = reproduced ? "Reproduced" : "Can't reproduce";
+            log.debug("{} with journal size: {} (took {}s)", outcome, m.getResult().size(), tookSeconds);
+            if (!reproduced) {
+                m.revert();
+            }
+        }
+        journal = new KieSessionJournalImpl(journal.getFacts(), journal.getInitialInserts(), m.getResult());
+        log.info("{} updates remaining.", journal.getMoveOperations().size());
+    }
+
+    private void pruneUpdates() {
+        log.info("Pruning updates...", journal.getMoveOperations().size());
+        RemoveRandomBlockMutator<KieSessionOperation> m = new RemoveRandomBlockMutator<KieSessionOperation>(journal.getMoveOperations());
+        while (m.canMutate()) {
+            log.debug("Current journal size: {}", m.getResult().size());
+            KieSessionJournalImpl testJournal = new KieSessionJournalImpl(journal.getFacts(), journal.getInitialInserts(), m.mutate());
+            boolean reproduced = reproduce(testJournal);
+            String outcome = reproduced ? "Reproduced" : "Can't reproduce";
+            List<KieSessionOperation> block = m.getRemovedBlock();
+            log.debug("{} without block of {} [{} - {}]",
+                      outcome, block.size(), block.get(0), block.get(block.size() - 1));
+            if (!reproduced) {
+                m.revert();
+            }
+        }
+        journal = new KieSessionJournalImpl(journal.getFacts(), journal.getInitialInserts(), m.getResult());
+        log.info("{} updates remaining.", journal.getMoveOperations().size());
+    }
+
+    private void pruneInserts() {
+        log.info("Pruning inserts...", journal.getInitialInserts().size());
+        RemoveRandomBlockMutator<KieSessionInsert> m = new RemoveRandomBlockMutator<KieSessionInsert>(journal.getInitialInserts());
+        while (m.canMutate()) {
```

Comment:
This could be extracted to separate method and called from both pruneUpdates() and pruneInserts().